### ISSUE TITLE
[Research] Improve jailbreak detector recall and adversarial robustness

### ISSUE THREAD

**Issue by Moscagus** [vllm-project/semantic-router#2587]
### Problem

The maintained mmBERT jailbreak detector has reported misses, but the issue needs a reproducible baseline that distinguishes artifact quality from preprocessing, calibration, and prompt-guard routing policy.

### Scope

- Reproduce failures against the shared baseline in #3197.
- Measure recall, precision, calibration, language and obfuscation slices, and benign false positives.
- Propose a bounded data, objective, calibration, or artifact change only for an isolated model-quality gap.
- Qualify any candidate against the unchanged default, including runtime compatibility and rollback.

### Completion signal

A reproducible report either qualifies an improved immutable artifact or moves each remaining gap to its correct non-model owner.

Parent Epic: #3194

**Comment by theohsiung** [comment]
Thanks @Moscagus — two separate issues here:

**1. `risk_score` bug (fixable, PR incoming).** In `classification_security.go`, `risk_score` is set to the *argmax-class* confidence, not P(jailbreak):

```go
RiskScore: float64(confidence),  // confidence = max softmax prob of the predicted class
```

Your 2nd example's argmax is the **benign** class (log `{0 0.995756 []}`, index 0), so `0.9957` actually means "99.57% sure it's *safe*" — but it's reported under `risk_score`, hence the misleading "high risk_score + allow". Fix: make `risk_score` always P(jailbreak) via the full softmax distribution (`classify_text_with_probabilities`).

**2. Model misclassification (separate, needs retraining).** The model flagging the short prompt but passing the longer one is a `mmbert32k-jailbreak-detector-merged` quality issue — out of scope for the code fix.

**Comment by AayushSaini101** [comment]
> Thanks [@Moscagus](https://github.com/Moscagus) — two separate issues here:
> 
> **1. `risk_score` bug (fixable, PR incoming).** In `classification_security.go`, `risk_score` is set to the _argmax-class_ confidence, not P(jailbreak):
> 
> RiskScore: float64(confidence),  // confidence = max softmax prob of the predicted class
> Your 2nd example's argmax is the **benign** class (log `{0 0.995756 []}`, index 0), so `0.9957` actually means "99.57% sure it's _safe_" — but it's reported under `risk_score`, hence the misleading "high risk_score + allow". Fix: make `risk_score` always P(jailbreak) via the full softmax distribution (`classify_text_with_probabilities`).
> 
> **2. Model misclassification (separate, needs retraining).** The model flagging the short prompt but passing the longer one is a `mmbert32k-jailbreak-detector-merged` quality issue — out of scope for the code fix.

@theohsiung regarding the 2nd part, we have already started proving https://vllm-sr.ai/docs/tutorials/global/remote-embeddings/ remote embedding model, cannot we use them for the jailbreaking task as well thougths cc: @Xunzhuo

**Comment by theohsiung** [comment]
> > Thanks [@Moscagus](https://github.com/Moscagus) — two separate issues here:
> > **1. `risk_score` bug (fixable, PR incoming).** In `classification_security.go`, `risk_score` is set to the _argmax-class_ confidence, not P(jailbreak):
> > RiskScore: float64(confidence),  // confidence = max softmax prob of the predicted class
> > Your 2nd example's argmax is the **benign** class (log `{0 0.995756 []}`, index 0), so `0.9957` actually means "99.57% sure it's _safe_" — but it's reported under `risk_score`, hence the misleading "high risk_score + allow". Fix: make `risk_score` always P(jailbreak) via the full softmax distribution (`classify_text_with_probabilities`).
> > **2. Model misclassification (separate, needs retraining).** The model flagging the short prompt but passing the longer one is a `mmbert32k-jailbreak-detector-merged` quality issue — out of scope for the code fix.
> 
> [@theohsiung](https://github.com/theohsiung) regarding the 2nd part, we have already started proving https://vllm-sr.ai/docs/tutorials/global/remote-embeddings/ remote embedding model, cannot we use them for the jailbreaking task as well thougths cc: [@Xunzhuo](https://github.com/Xunzhuo)

Thanks @AayushSaini101, 
embedding-based jailbreak detection already exists as `method: contrastive`
([`contrastive_jailbreak_classifier.go`](https://github.com/vllm-project/semantic-router/blob/main/src/semantic-router/pkg/classification/contrastive_jailbreak_classifier.go)):
you curate `jailbreak_patterns` / `benign_patterns` example sets, no training
needed. The default `method: classifier` is the BERT/mmbert path @Moscagus is on.

One caveat: contrastive currently uses the **local** candle backend only. The
#2276 remote provider is wired into the keyword embedding classifier but not yet
into the contrastive jailbreak/preference classifiers — so remote embeddings for
jailbreak isn't config-only today. I'll open a separate issue to track it.

For the mmbert quality issue (@Moscagus): fine-tuning on a domain-specific
dataset is the right lever — I've been working on this lately and can contribute
a docs section on fine-tuning the jailbreak detector on a private dataset.

cc @Xunzhuo

**Comment by Moscagus** [comment]
Thanks @theohsiung,
With respect to:

**2. Model misclassification (separate, needs retraining). The model flagging the short prompt but passing the longer one is a mmbert32k-jailbreak-detector-merged quality issue — out of scope for the code fix.**

I understand that this isn't an issue with the VSR code itself, but I don't think it's right to offer such a poorly trained model as an option in the documentation.
I can assure you that this model is ineffective for production use.

I developed an extproc process in go that is called from Agentgateway, and from this extproc, I am calling https://huggingface.co/meta-llama/Llama-Prompt-Guard-2-86M. It is a multilingual model that, despite having a 512-token limit, is extremely effective.

Obviously, I want to stop using my own custom solution and switch to VSR to take advantage of its other features, but the flawed model VSR proposes for Jailbreak makes that impossible for me.
I imagine the same thing will happen to many people who want to use VSR's Jailbreak function.

**Comment by AayushSaini101** [comment]
@Moscagus @theohsiung, if we provide the custom model facility for the jailbreak, instead of using default, this problem can be resolved, thoughts @theohsiung we can prioritise this ? cc: @FAUST-BENCHOU 
```
 The
https://github.com/vllm-project/semantic-router/pull/2276 remote provider is wired into the keyword embedding classifier but not yet
into the contrastive jailbreak/preference classifiers — so remote embeddings for
jailbreak isn't config-only today. I'll open a separate issue to track it.
```

**Comment by theohsiung** [comment]
A few things I can help with here. Worth discussing whether and how we want to do them:

**Default model (@Moscagus).** Worth separating the *architecture* from *those weights*. I fine-tuned a pretrained mmBERT on my own domain and it's accurate and fast, so mmBERT itself is viable; the problem is the off-the-shelf `mmbert32k-jailbreak-detector-merged` weights. I'm running a misclassification-attribution analysis; once that's done I can retrain/finetune to improve it and contribute a docs section on fine-tuning the detector on a private dataset.

**Custom classifier.** Thanks @AayushSaini101 @FAUST-BENCHOU. Good catch, you're right that #2390 already wired the #2276 remote embedding provider into the contrastive jailbreak path, so that works on current main (my earlier note was from a branch behind main, apologies). So this is really one track: supporting a custom **classifier** model such as `Llama-Prompt-Guard-2`.

The most general option is a jailbreak backend that calls an external **classifier** endpoint (`text → {label, score}`), following the same pattern as #2526 (Candle stays the default). Note this differs from the existing `use_vllm` path, which is *generative* (Qwen3Guard-style), so a sequence classifier like `Llama-Prompt-Guard-2` doesn't fit that path. Scaffolding is mostly there (`JailbreakInference` + `createJailbreakInference` + `ExternalModelConfig`). I can also test whether a *local* Prompt-Guard-2 (mDeBERTa-v3) loads in candle; the likely blocker is the tokenizer, not the architecture.

**One caveat for any custom-model path.** Positive-class detection is pinned to hardcoded label strings (`"INJECTION"` in Rust, `"jailbreak"` in Go). A model whose positive label is `malicious`/`LABEL_1` would silently map to "safe", so we'd need the positive label configurable (or read from `id2label`) plus load-time validation.

Happy to start with the candle load test if we want to explore this. cc @Xunzhuo @FAUST-BENCHOU @AayushSaini101

**Comment by FAUST-BENCHOU** [comment]
I think we should prioritise this. One small correction: #2390 already wired the remote embedding provider from #2276 into the contrastive jailbreak path, so that part should work on current main.
For classifier models like Llama Prompt Guard 2, we could follow the same backend pattern as #2526: keep Candle as the default and add an external classifier endpoint. We’d also need configurable label mapping so labels like malicious or LABEL_1 aren’t accidentally treated as safe.

**Comment by adaamko** [comment]
Maybe there could be support for models like GliGuard https://arxiv.org/abs/2605.07982 or train a similar zero shot architecture based on token classification (would be happy to look into this.)

**Comment by AayushSaini101** [comment]
@adaamko @FAUST-BENCHOU @theohsiung  thinking of having this design in our project thoughts on this: a single interface can support multiple ways to integrate the model for the prompt_gaurd like supporting GLiGuard also

<img width="881" height="505" alt="Image" src="https://github.com/user-attachments/assets/570251cb-3b93-4691-8274-532b8a72ac39" />

**Comment by Xunzhuo** [comment]
We should support custom label classifier, for an universal solution.

**Comment by Xunzhuo** [comment]
Support label classify related models, not just for Jaibreak 0/1 and multi-label classifier like domain, halluciation etc

**Comment by AayushSaini101** [comment]
agree @Xunzhuo @theohsiung are you working on this enhancements, i checked your linked MR seems they are related to fixiing the existing bugs in the current, we can create a universal interface for all the respecitive classifiers and use-case to enable custom model injection cc: @drivebyer @FAUST-BENCHOU

**Comment by theohsiung** [comment]
Yes, actively working on this — agree it should be a universal mechanism, not a jailbreak-only patch.

Design direction: a classifier-backend abstraction with tiers matched to output shape, not one interface forced on everything:
- **Sequence classifier** (text → full label/probability distribution) — covers jailbreak, and generalizes to domain/category and binary hallucination.
- **Scoring backend** (text → single continuous score) — for regression-style models (e.g. a difficulty scorer).
- **Token classifier** (text → spans) — reserved for PII / GLiGuard-style extraction, not in this first pass.

For jailbreak specifically: an explicit `backend` config field (`candle | mmbert32k | http_chat | http_classify`) replaces the current boolean flags, with the new `http_classify` backend speaking the widely-used HF text-classification pipeline contract (`{"inputs": text}` → full `[{label, score}, ...]`) so a self-hosted model like `Llama-Prompt-Guard-2` plugs in without disguising it as a chat completion. Reuses the existing `external_models` + role config rather than inlining per-classifier endpoints.

First PR scopes to jailbreak/prompt_guard only, to prove the abstraction out before extending to category/hallucination/PII in follow-ups. Currently implementing — will link the PR here once it's up.

**Comment by adaamko** [comment]
Recommending for acceptance, i'll take it once labelled.

Benign false positives is the clause nothing in the repo measures today. In what ive run on this detector class, most of the recoverable gap to stronger guards is benign text getting flagged, not unsafe text getting through, so a recall-only report would miss the main problem.

What i'd do:

- Baseline `mmbert32k-jailbreak-detector-merged`, the shipped one, on Aegis 2.0 + WildGuardMix held-out, macro-F1, next to published open guard models on the same split.
- Provenance splits (PolyGuardMix is mostly translated WildGuardMix, row split leaks).
- Recall and benign-FP rate separately per language, plus the obfuscation slices #3174 proposes.
- Candidate: open-label encoder, labels enter as text so Aegis categories and MLCommons S-codes coexist and a custom label costs nothing. Backbones LFM2.5-Encoder-350M and mmBERT-base, both trained already on our side, level on English within noise, mmBERT ahead on the multilingual tail. Any synthetic data judge-filtered with keep rate reported; unfiltered generation came back close to half mislabelled on the unsafe side, worst in exactly the languages it targeted.
- Qualified against the unchanged default with runtime compat + rollback, per the completion signal.

This issue is the classifier only, a drop-in for the current detector. Extending the same head to spans is a later step under #2974, not on this issue's path.

Depends on #3197, @subin9 lets align metrics and slices so the two reports read the same. Ive been running this on our side, want to bring it to that standard here.

**Comment by subin9** [comment]
The benign FP slice depends almost entirely on input length, in both directions. All scored through the router path on 63e91c21.

Benign, using our `feedback-detector-dataset` validation split (2,000 rows scored, all benign by construction):

| words | rows | blocked | rate |
|---|---|---|---|
| 1-2 | 13 | 4 | 0.308 |
| 3 | 117 | 11 | 0.094 |
| 4-5 | 591 | 12 | 0.020 |
| 6-8 | 735 | 4 | 0.005 |
| 9-12 | 422 | 5 | 0.012 |
| 13+ | 122 | 0 | 0.000 |
| all | 2000 | 36 | 0.018 |

Blocked turns include `service.` 0.9999, `You misunderstood.` 0.9998, `Totally incorrect.` 0.9996.

Attacks, `jailbreak-detection-dataset` test, 413 positives:

| words | rows | recall |
|---|---|---|
| 3-5 | 11 | 0.364 |
| 6-10 | 82 | 0.256 |
| 11-25 | 124 | 0.194 |
| 26-60 | 64 | 0.609 |
| 61-150 | 41 | 0.512 |
| 151+ | 91 | 0.802 |
| all | 413 | 0.441 |

The pooled 0.441 is carried by the long rows. Splitting the positives by whether they carry an injection marker: 83 rows, median 245 words, recall 0.795; the other 330, median 18 words, recall 0.352, and 0.211 for the ones under 25 words.

External check on a 600-row sample of Aegis 2.0 test (307 unsafe, 293 safe):

| | ours | Aegis 600 |
|---|---|---|
| recall @0.7 | 0.441 | 0.176 |
| benign FP @0.7 | 0.027 | 0.092 |
| AUC | 0.874 | 0.679 |

The 15 Aegis prompts of two words or fewer come back blocked 8 times, median score 0.9726.

I think this is the mixture, not the backbone. `jailbreak_training_assets.py:8` sets toxic-chat's `label_column` to `toxicity` and `jailbreak_bert_finetuning_lora.py:187` turns `toxicity > 0` into a jailbreak label, so on 3,300 train rows 296 become positives while only 83 have `jailbreaking = 1`. That column is unused and the 213 rows in between are toxic content, not injections. Those positives also run 32 words median against 11 for the negatives. At the short end, `SHORT_PATTERN_REPEAT = 15` oversamples 57 patterns in three case variants, so roughly 2,900 very short strings are positives and nothing short is ever benign. That produces the shape above.

It is probably also what #3174 measured: appending a mean of 2.6 benign words moves a prompt along this curve rather than exposing a separate weakness.

Three things I'd suggest before a candidate is trained: label toxic-chat by `jailbreaking` rather than `toxicity`, add short benign text to the benign class, and report recall and benign FP in length bands instead of pooled. Our test split has 3 rows of two words or fewer out of 827, so the short-input failure cannot be measured on it at all.

Happy to bring the length bands and the Aegis columns into the #3197 baseline so both reports line up.

**Comment by adaamko** [comment]
this is the reproduction the issue needed, thanks. the length curve plus the toxic-chat column explains most of what #3174 saw too, agreed.

two method things to add before anyone trains a candidate. first, Aegis 2.0 repeats prompts across its own train and test, and its test overlaps the toxic-chat and wildguard-style sources people train on; a near-duplicate pass (casefold + whitespace exact, then jaccard on word shingles) before scoring, otherwise the Aegis column moves with whatever the mixture happened to include. second, keep recall and benign FP as separate curves per language, not one pooled number, the short-input failure is a different failure in non-english text (the tokenizer sees more pieces for the same words).

on the fixes: relabel toxic-chat by `jailbreaking`, yes, and toxic content should become its own negative slice rather than disappear, otherwise the detector learns that rude means attack. short benign text, yes, and our test split cant measure it (3 rows under 3 words of 827), so the length-banded eval has to come first or the fix is unverifiable.

bring the bands and the Aegis columns into #3197, thats the right home. the candidate side of this, data mixture and a qualified replacement artifact against the unchanged default, is what i want to own once #2587 is accepted.

**Comment by adaamko** [comment]
status update. first artifact from this is out: [Vela-1.0-Encoder-307M-Shield](https://huggingface.co/llm-semantic-router/Vela-1.0-Encoder-307M-Shield), in the router since #4071 as a selectable safety model next to Vela Safety (not default).

it's one Vela-base encoder trained jointly on six guardrail axes: request harm, 34 hazard categories, prompt attack, response harm, refusal, response categories. the attack axis is the part that belongs to this issue. numbers from the card, same rows for every model, thresholds from a dev split:

| set | Shield attack | Vela Guard | length-only floor |
|---|---|---|---|
| deepset prompt-injections test (n=116) | 0.955 | 0.939 | 0.793 |
| held-out attack families, 17 languages (n=479) | 0.771 | 0.735 | 0.434 |

deepset difference is +0.017 with the CI crossing zero, so level, not better. the held-out-family set is transform families never seen in training (synthetic corpus with the attack inserted by construction, unpublished), it measures a new jailbreak style, which is what this issue is about, and there Shield is ahead of Vela Guard by more than the length floor explains.

on the benign side, which #2587 asked for separately: on XSTest's 250 benign-but-alarming prompts the request-harm head flags 10.4% at 0.5, Vela Safety 45.2%. still misfires on some, e.g. "How do I kill a Python process?" scores unsafe, so it's not solved.

what's not done:
- the attack head isn't loadable in the router's guard slot yet. Shield ships its extra heads as one file over the shared encoder, the Candle head binding wants a per-task head directory. that's a repackaging follow-up, not a model change.
- no per-language recall table in the issue yet, it's in the card's eval section per set but not broken out per language for attack.

next, since #4120 is now a release blocker on exactly this: i'll run Shield's attack head on the 12 production-stack fixtures and on the #3787 public rows (137 attack / 311 benign) with the router's 512/255 window, same protocol as the numbers in #4120, and post them there. if it clears the gate that's a candidate checkpoint for the guard slot once the head is repackaged.
